### PR TITLE
Support grouped execution for window function

### PR DESCRIPTION
**Pull request by shixuan-fan**
Depends on #12122

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 1]
```diff
@@ -2723,6 +2729,48 @@ public void testGroupedExecution()
...
+
+            assertQuery(
+                    colocatedOneGroupAtATime,
+                    "SELECT row_number() OVER (PARTITION BY key) FROM test_window",
+                    expectedWindowFunctionResult,
+                    assertRemoteExchangesCount(1));
+
+            assertQuery(
+                    colocatedOneGroupAtATime,
+                    "SELECT row_number FROM (SELECT row_number() OVER (PARTITION BY key ORDER BY value) as row_number FROM test_window) WHERE row_number = 1",
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Change `row_number` to `n` in all of these to make the test query shorter. Also capitalize `AS` since it's a keyword (or remove it since it is optional).

Same comment for the similar query below.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 2]
```diff
@@ -596,6 +599,36 @@ public GroupedExecutionProperties visitAggregation(AggregationNode node, Void co
...
+
+        @Override
+        public GroupedExecutionProperties visitTopNRowNumber(TopNRowNumberNode node, Void context)
+        {
+            return visitWindowFunctionNodeInternal(node);
+        }
+
+        private GroupedExecutionProperties visitWindowFunctionNodeInternal(PlanNode node)
+        {
+            checkState(
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: This should be `checkArgument`, since it is validating the method argument.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java [hunk 3]
```diff
@@ -596,6 +599,36 @@ public GroupedExecutionProperties visitAggregation(AggregationNode node, Void co
...
+                    node instanceof WindowNode || node instanceof RowNumberNode || node instanceof TopNRowNumberNode,
+                    "Expects only WindowNode, RowNumberNode or TopNRowNumberNode");
+            GroupedExecutionProperties properties = node.getSources().get(0).accept(this, null);
+            if (properties.isCurrentNodeCapable()) {
+                return new GroupedExecutionProperties(true, true, properties.capableTableScanNodes);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: Use the getter rather than accessing the private field directly.

Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java]: I found the usage of getter is not consistent in this class. Would you suggest a refactor commit to make them consistent before the current commit?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 4]
```diff
@@ -2723,6 +2729,48 @@ public void testGroupedExecution()
...
             assertQuery(colocatedOneGroupAtATimeDynamic, chainedSharedBuildOuterJoin, expectedChainedOuterJoinResult, assertRemoteExchangesCount(2));
 
+            //
+            // Window function
+            // ===============
+            // Window partition key covers bucket key
+            @Language("SQL") String expectedWindowFunctionResult = format("VALUES %s", String.join(", ", nCopies(15000, "(1)")));
+            assertQuery(
+                    colocatedOneGroupAtATime,
+                    "SELECT rank() OVER (PARTITION BY key) FROM test_window",
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: These tests looks quite trivial since they always partition on a column with unique values. 

Can we see if there are any test queries for window function that we can use here?

It looks like for other use case (JOIN/AGGREGATION), we never don't test the situation where it's not suitable for grouped execution (supposedly, they are covered by other integration tests). So I guess we don't need the tests `Window partition key does not cover bucket key`.

**Comment by shixuan-fan** [comment]
Comment addressed except `checkArgument` one (https://github.com/prestodb/presto/pull/12234#discussion_r248907469). There was discussion to remove it.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 5]
```diff
@@ -2723,6 +2731,37 @@ public void testGroupedExecution()
...
+            //
+            // Window function
+            // ===============
+            ImmutableList.Builder<String> expectedCountValues = ImmutableList.builder();
+            expectedCountValues.addAll(nCopies(7333, "('O', 7333)"));
+            expectedCountValues.addAll(nCopies(7304, "('F', 7304)"));
+            expectedCountValues.addAll(nCopies(363, "('P', 363)"));
+            @Language("SQL") String expectedCountResult = format("VALUES %s", String.join(", ", expectedCountValues.build()));
+            assertQuery(
+                    colocatedOneGroupAtATime,
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: I think we should rename this variable to some thing like `groupedExecutionOneGroupAtATime`. But we can do this in separate PR.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 6]
```diff
@@ -2369,6 +2372,11 @@ public void testGroupedExecution()
                             "WITH (bucket_count = 13, bucketed_by = ARRAY['keyD']) AS\n" +
                             "SELECT orderkey keyD, comment valueD FROM orders CROSS JOIN UNNEST(repeat(NULL, 2))",
                     30000);
+            assertUpdate(
+                    "CREATE TABLE test_window\n" +
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: nit: Maybe call it `test_grouped_window`? 

Also remember to use 13 buckets. Since 13 is co-prime with 4 (number of tasks), thus it tests unbalanced assignment ;)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 7]
```diff
@@ -2723,6 +2731,37 @@ public void testGroupedExecution()
             assertQuery(colocatedOneGroupAtATime, chainedSharedBuildOuterJoin, expectedChainedOuterJoinResult, assertRemoteExchangesCount(2));
             assertQuery(colocatedOneGroupAtATimeDynamic, chainedSharedBuildOuterJoin, expectedChainedOuterJoinResult, assertRemoteExchangesCount(2));
 
+            //
+            // Window function
+            // ===============
+            ImmutableList.Builder<String> expectedCountValues = ImmutableList.builder();
+            expectedCountValues.addAll(nCopies(7333, "('O', 7333)"));
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Maybe add some filter when create the `test_grouped_window` table? Enumerate 7333 tuples in `VALUES` sounds a bit too much. Maybe make about 10% of the tuples?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 8]
```diff
@@ -2766,9 +2805,17 @@ public void testGroupedExecution()
             assertUpdate("DROP TABLE IF EXISTS test_grouped_join3");
             assertUpdate("DROP TABLE IF EXISTS test_grouped_joinN");
             assertUpdate("DROP TABLE IF EXISTS test_grouped_joinDual");
+            assertUpdate("DROP TABLE IF EXISTS test_window");
         }
     }
 
+    private List<String> getRowNumberValues(String value, int maxRowNumber)
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: nit: static

Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: nit: static

**Comment by shixuan-fan** [comment]
As discussed offline with @wenleix , we will limit the test to 10 entries and enumerate the result, instead of populating values by code, to make it more readable.

**Comment by shixuan-fan** [comment]
Test updated.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 9]
```diff
@@ -2735,6 +2740,47 @@ public void testGroupedExecution()
...
+                            "(4, 3),\n" +
+                            "(4, 4),\n" +
+                            "(4, 5),\n" +
+                            "(5, 1),\n" +
+                            "(5, 2)",
+                    assertRemoteExchangesCount(1));
+
+            assertQuery(
+                    colocatedOneGroupAtATime,
+                    "SELECT n FROM (SELECT row_number() OVER (PARTITION BY key ORDER BY value) AS n FROM test_grouped_window) WHERE n = 1",
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: I assume this is trying to test row number top n. Why only select `row_number()` but ignore `key`? 

Also does it makes sense to do something like where `n <= 2` so the test is not so trivial>

**Comment by wenleix** [comment]
See also: https://github.com/prestosql/presto/pull/169
